File: apps/web/src/lib/terminal/messageProtocol.ts
export type ClientMessageType =
  | 'create_session'
  | 'input'
  | 'resize'
  | 'close_session';

export type ServerMessageType =
  | 'output'
  | 'session_created'
  | 'session_closed'
  | 'error';

export interface CreateSessionMessage {
  type: 'create_session';
  sessionId: string;
}

export interface InputMessage {
  type: 'input';
  sessionId: string;
  data: string;
}

export interface ResizeMessage {
  type: 'resize';
  sessionId: string;
  cols: number;
  rows: number;
}

export interface CloseSessionMessage {
  type: 'close_session';
  sessionId: string;
}

export type ClientMessage =
  | CreateSessionMessage
  | InputMessage
  | ResizeMessage
  | CloseSessionMessage;

export interface OutputMessage {
  type: 'output';
  sessionId: string;
  data: string;
}

export interface SessionCreatedMessage {
  type: 'session_created';
  sessionId: string;
}

export interface SessionClosedMessage {
  type: 'session_closed';
  sessionId: string;
}

export interface ErrorMessage {
  type: 'error';
  sessionId: string;
  error: string;
}

export type ServerMessage =
  | OutputMessage
  | SessionCreatedMessage
  | SessionClosedMessage
  | ErrorMessage;

File: apps/api/services/message_protocol.py
from typing import Literal, TypedDict


class CreateSessionMessage(TypedDict):
    type: Literal["create_session"]
    sessionId: str


class InputMessage(TypedDict):
    type: Literal["input"]
    sessionId: str
    data: str


class ResizeMessage(TypedDict):
    type: Literal["resize"]
    sessionId: str
    cols: int
    rows: int


class CloseSessionMessage(TypedDict):
    type: Literal["close_session"]
    sessionId: str


ClientMessage = CreateSessionMessage | InputMessage | ResizeMessage | CloseSessionMessage


class OutputMessage(TypedDict):
    type: Literal["output"]
    sessionId: str
    data: str


class SessionCreatedMessage(TypedDict):
    type: Literal["session_created"]
    sessionId: str


class SessionClosedMessage(TypedDict):
    type: Literal["session_closed"]
    sessionId: str


class ErrorMessage(TypedDict):
    type: Literal["error"]
    sessionId: str
    error: str


ServerMessage = OutputMessage | SessionCreatedMessage | SessionClosedMessage | ErrorMessage
